### ISSUE TITLE
node fromPNG toPNG and generic fromPixels toPixels

### ISSUE THREAD

**Issue by DavidGOrtega** [tensorflow/tfjs#764]
I won't use fromPixels and toPixels in node since It uses Cairo and I'm focused on Windows and Mac. Cairo on Windows apart of having to maintain additional binaries it's not as easy to install as other OS.

If you are like me that want to keep the binaries to the bare minimum here is a very fast fromPNG and toPNG for node purely using js.

```
npm install pngjs
```
```js
const fs      = require('fs'),
const PNG  = require('pngjs').PNG; 

async function fromPNG(path, numChannels)
{
    numChannels = numChannels || 3;
    
    return new Promise((resolve, reject) => 
    {
        fs.createReadStream(path).pipe(
        new PNG({colorType: 6} ))
        .on('parsed', function() 
        {
            resolve( fromPixels(this.data, this.width, this.height, numChannels) );
        })
        .on('error', reject);
    });
}

function toPNG(tensor, path)
{
    let dst     = new PNG({width: tensor.shape[0], height: tensor.shape[1]});
    dst.data    = toPixels(tensor);
    dst.pack().pipe( fs.createWriteStream(path) );
}

function fromPixels(data, width, height, numChannels)
{
    const values    = [];
    const numPix    = width * height;
    
    for (let i = 0; i < numPix; i++) 
        for (let channel = 0; channel < numChannels; ++channel)
            values[i * numChannels + channel] = data[i * 4 + channel];

    let tensor = tf.tensor3d(values, [width, height, numChannels], 'int32');
    
    return tensor;
}

function toPixels(tensor) 
{
    const pixels    = tensor.dataSync();
    const data      = [];
    
    if (tensor.shape.length == 2 || tensor.shape[2] == 1) 
    {
        // Grayscale
        for (let i = 0; i < pixels.length; i++) {
          data[i*4+0] = pixels[i];
          data[i*4+1] = pixels[i];
          data[i*4+2] = pixels[i];
          data[i*4+3] = 255;
        }
        
    } else if (tensor.shape[2] == 3) 
    {
        // RGB
        for (let i = 0; i < pixels.length / 3; i++) 
        {
          data[i*4+0] = pixels[i*3+0];
          data[i*4+1] = pixels[i*3+1];
          data[i*4+2] = pixels[i*3+2];
          data[i*4+3] = 255;
        }
        
    }else if (tensor.shape[2] == 4) 
    {
        // RGBA
        data = pixels;
    }
    
    return data;
}
```

JPG could be the same using node-jpg. Of course like fromPixels or toPixels assumes that your shapes are going to be [W, H, Channels] maybe you have to swap them if your shapes are [Channels, W, H].
Also you have data on the range [0, 255] so don't forget to preprocess and postprocess to fit your activations.

**Comment by davidsoergel** [comment]
@DavidGOrtega would you like to send a PR about this?  @nkreeger WDYT?

**Comment by nkreeger** [comment]
@davidsoergel @DavidGOrtega cool nice work - we're going to be shipping the generated tf.fromPixels() API soon. That's the official one we want to support since it uses the TF C-library to convert JPEG/PNG/etc byte streams to Tensors in C-land.

**Comment by davidsoergel** [comment]
Closing based on @nkreeger's comment.  Thanks for the helpful feedback & code, though!

**Comment by oveddan** [comment]
@nkreeger is this api method avail?  Would be great to use it!  Also @DavidGOrtega there is an issue with your code - it should always be height first in the dimensions of tensors (not width first as in your code).

**Comment by kangyizhang** [comment]
@oveddan the api is available in head and you can try it out by [publish tfjs-node locally](https://github.com/tensorflow/tfjs-node/blob/master/DEVELOPMENT.md#publish-locally-through-yalc-to-test-this-package-in-another-repo) and install it through yalc. We'll publish a new version soon which would include the api.

**Comment by oveddan** [comment]
Thanks @kangyizhang I scouted out the code and this is super cool, esp how its tied to the c-library!
Unfortunately I've already been through a world of pain to get tfjs-node-gpu working on my windows pc, and now when I try to do the aformentioned steps:
(in tfjs-node dir):
```
yarn
yarn enable-gpu 
yarn publish-local
```

in my project directory:
`yalc link @tensorflow/tfjs-node`

Everything works but then when I run the program:
Error: Cannot find module 'C:\Users\Al\Source\pose-extraction\.yalc\@tensorflow\tfjs-node\lib\napi-v4\tfjs_binding.node'

I think I'll just wait for the release to come out.

**Comment by kangyizhang** [comment]
@oveddan FYI tfjs-node@1.2.5 has been released.

**Comment by DavidGOrtega** [comment]
> JPG could be the same using node-jpg. Of course like fromPixels or toPixels assumes that your shapes are going to be [W, H, Channels] maybe you have to swap them if your shapes are [Channels, W, H].
Also you have data on the range [0, 255] so don't forget to preprocess and postprocess to fit your activations.

@oveddan depends on the model. As I stated in my comment there are so many ways. That code actually works for an encoder-decoder model that I have. Was just only a recipe not a universal solution. I did it just because I found my self very limited in node js managing images.
